metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/BlockSynchronizer.scala

```diff
@@ -0,0 +1,218 @@
+  * a long time offline. Once that happens the block history should be pruneable.
+  */
+class BlockSynchronizer[F[_]: Sync: Timer, N, A <: Agreement: Block](
+    blockStorage: BlockStorage[N, A],
+    getBlock: BlockSynchronizer.GetBlock[F, A],
+    inMemoryStore: KVStoreRunner[F, N],
+    semaphore: Semaphore[F],
+    retryTimeout: FiniteDuration = 5.seconds
+)(implicit storeRunner: KVStoreRunner[F, N]) {
+
+  // We must take care not to insert blocks into storage and risk losing
+  // the pointer to them in a restart, hence keeping the unfinished tree
+  // in memory until we find a parent we do have in storage, then
+  // insert them in the opposite order.
+
+  /** Download all blocks up to the one included in the Quorum Certificate.
+    *
+    * Only expected to be called once per sender at the same time, otherwise
+    * it may request the same ancestor block multiple times concurrently.
+    *
+    * This could be managed with internal queueing, but not having that should
+    * make it easier to cancel all calling fibers and discard the synchronizer
+    * instance and its in-memory store, do state syncing, then replace it with
+    * a fresh one.
+    */
+  def sync(
+      sender: A#PKey,
+      quorumCertificate: QuorumCertificate[A]
+  ): F[Unit] =
+    for {
+      path <- download(sender, quorumCertificate.blockHash, Nil)
+      _    <- persist(quorumCertificate.blockHash, path)
+    } yield ()
+
+  /** Download a block and all of its ancestors into the in-memory block store.
+    *
+    * Returns the path from the greatest ancestor that had to be downloaded
+    * to the originally requested block, so that we can persist them in that order.
+    *
+    * The path is maintained separately from the in-memory store in case another
+    * ongoing download would re-insert something on a path already partially removed
+    * resulting in a forest that cannot be traversed fully.
+    */
+  private def download(
+      from: A#PKey,
+      blockHash: A#Hash,
+      path: List[A#Hash]
+  ): F[List[A#Hash]] = {
+    storeRunner
+      .runReadOnly {
+        blockStorage.contains(blockHash)
+      }
+      .flatMap {
+        case true =>
+          path.pure[F]
+
+        case false =>
+          inMemoryStore
+            .runReadOnly {
+              blockStorage.get(blockHash)
+            }
+            .flatMap {
+              case Some(block) =>
+                downloadParent(from, block, path)
+
+              case None =>
+                getAndValidateBlock(from, blockHash)
+                  .flatMap {
+                    case Some(block) =>
+                      inMemoryStore.runReadWrite {
+                        blockStorage.put(block)
+                      } >> downloadParent(from, block, path)
+
+                    case None =>
+                      Timer[F].sleep(retryTimeout) >>
+                        download(from, blockHash, path)
```

Comment:
Currently `sync` is only called in response to `prepare` message and we are trying to sync from only the sender (leader). If the sender becomes unresponsive this will loop forever. If we sync from the replicas as well (as you mention in another comment), then it should resolve this (someone should ultimately respond or it's us who are the problem). So when are we going to sync from the replicas?

Reply:
Yeah, I didn't want to complicate it with a download manager that has a queue and tracks the available sources, so at the moment the way it works is that we only ever try to download blocks that have a Quorum Certificate, so they are not completely rubbish, nobody can take us down the garden path with a block that has parents that don't lead anywhere. But I quite possibly missed something.

Say Alice is the leader this round and she sends us block B7 to prepare, but for whatever reason we are missing B5, so we start downloading that from Alice. My assumption was that if for whatever reason this keeps failing, the next leader, Bob, will propose something else built on B5, and we'll just get it from him.

But if he's using a different prepare quorum certificate then indeed B5 may never again be build upon and we might never get it from Alice. 

I guess it would be prudent to add some clause to give up after a while and just respond that the block proposed by Alice is not valid, or never even respond to that in the `SyncPipe`.

Reply:
Hm, unfortunately the `BlockSynchronizer` is a bit simplistic in this regard: because we have already inserted whatever we _could_ download from Alice into the in-memory tree, and we don't know what else might depend on them, we can't just remove them on our way out, because it could clash with the `unpersisted` mechanism described above. 

Our best option is to ask another federation member.

Reply:
Should we add a TODO comment to reconsider it later?

Reply:
Pushed a change that asks the others in a randomized order. That should not fail, but if it does, it is going to retry forever until somebody serves the block.

Reply:
Interesting, I like it :+1: